// zombiecss/src/zombiecss.js
var fs = require('fs')
var Mustache = require('mustache')
var resemble = require('node-resemble')
var svg2png = require("svg2png")
var base64Img = require('base64-img')

function injectScript(zombie, file) {
    var script = zombie.document.createElement('script')
    script.setAttribute('src', 'file://' + file)
    script.setAttribute('type', 'text/javascript')
    zombie.document.body.appendChild(script)
}

function runJs(zombie, javascriptString) {
    var script = zombie.document.createElement('script')
    script.setAttribute('type', 'text/javascript')
    script.innerHTML = javascriptString
    zombie.document.body.appendChild(script)
}

function saveDomToSvg(file, dom, style, links, width, height, toSvg) {
    var svg = Mustache.render(fs.readFileSync(__dirname + '/../templates/svg.svg').toString(), {
        links: links,
        style: style,
        dom: dom,
        width: width,
        height: height
    });

    if (toSvg) {
        fs.writeFileSync(file + '.svg', svg)
    } else {
        var buffer = new Buffer(svg)
        fs.writeFileSync(file + '.png', svg2png.sync(buffer))
    }
}

function screenshot(zombie, options) {
    options = options || {}
    options.width = options.width || 10000
    options.height = options.height || 10000

    var allStyle = ''
    var styles = zombie.document.getElementsByTagName('style')
    Object.keys(styles).forEach(function(index) {
        var style = styles[index]
        allStyle += style.innerHTML + '\n'
    })

    var allLinks = ''
    var links = zombie.document.getElementsByTagName('link')
    Object.keys(links).forEach(function(index) {
        var link = links[index]
        allLinks += Mustache.render(fs.readFileSync(__dirname + '/../templates/link.html').toString(), {
            href: link.href
        }) + '\n';
    })

    zombie.wait().then(function() {
        saveDomToSvg(options.baseDir + '/' + options.actualFile, zombie.document.querySelector(options.selector).outerHTML, allStyle, allLinks, options.width, options.height, options.toSvg)
    })
}

function compare(zombie, options) {
    options.toPng = true
    screenshot(zombie, options)

    zombie.wait().then(function() {
        resemble(options.baseDir + '/' + options.actualFile + '.png')
            .compareTo(options.baseDir + '/' + options.expectedFile + '.png')
            .onComplete(function(data) {
                console.log(data) //save the rest of the json
                base64Img.imgSync(data.getImageDataUrl(), options.baseDir, options.diffFile)
            })
    })
}

module.exports = {
    screenshot: screenshot,
    compare: compare
}
